Snuba tagstore backend

alex-hofsteede opened the pull request.

alex-hofsteede committed:
Simple snuba version of get_group_tag_keys_and_top_values

alex-hofsteede committed:
Top values for tag.

alex-hofsteede committed:
Move snuba tagstore tests into snuba test suite

alex-hofsteede committed:
Get tag count

alex-hofsteede committed:
Get group(s) tag value

alex-hofsteede committed:
Get tag key/keys

alex-hofsteede committed:
Add methods for tag/value lookup without group

alex-hofsteede committed:
Get groups user counts

alex-hofsteede committed:
Get first and last releases

alex-hofsteede committed:
Get group ids for users

alex-hofsteede committed:
Get release tags

alex-hofsteede committed:
Get group event ids

alex-hofsteede committed:
Get group tag values for users

bretthoerner commented on src/sentry/tagstore/snuba/backend.py:
```diff
@@ -16,69 +17,363 @@
+        """
+        Returns the default (start, end) time range for querrying snuba.
+        """
+        end = datetime.utcnow().replace(tzinfo=pytz.UTC)
```
nit: I noticed in search we seem to use `django.utils.timezone.now` instead of `datetime.utcnow().replace(tzinfo=pytz.UTC)`. It apparently does the same thing but respects some Django setting?

bretthoerner commented on src/sentry/tagstore/snuba/backend.py:
```diff
@@ -16,69 +17,363 @@
     # Tag keys and values
     def get_tag_key(self, project_id, environment_id, key, status=TagKeyStatus.VISIBLE):
-        pass
+        # TODO is status applicable?
```
I don't think so, it's only applicable for things in the progress of going through a (sometimes cascading) async delete.

bretthoerner commented on src/sentry/tagstore/snuba/backend.py:
```diff
@@ -16,69 +17,363 @@
 
     def get_tag_keys(self, project_id, environment_id, status=TagKeyStatus.VISIBLE):
-        pass
+        # TODO why is there no limit?
```
I think the assumption (wrong as it may be) in existing callers of this code is that no single project/group would have so many TagKeys that it would matter. I think this is generally true unless the user is sending dynamically generated tag names that are unique, which is pretty much always an error I believe.

bretthoerner commented on src/sentry/tagstore/snuba/backend.py:
```diff
@@ -16,69 +17,363 @@
 
 class SnubaTagStorage(TagStorage):
 
+    def get_time_range(self, days=90):
```
We probably want to note a TODO to use something like this: https://github.com/getsentry/sentry/blob/9c87c1ae59f6d1eae39e2bcecee0b192cb8bf3f5/src/sentry/search/django/backend.py#L241-L251

alex-hofsteede replied:
done

bretthoerner commented on src/sentry/tagstore/snuba/backend.py:
```diff
@@ -16,69 +17,363 @@
     def get_tag_key(self, project_id, environment_id, key, status=TagKeyStatus.VISIBLE):
-        pass
+        # TODO is status applicable?
+        return self.get_group_tag_key(project_id, None, environment_id, key)
```
I didn't look up where we depend on this again but we may need to wrap this to catch `GroupTagKeyNotFound` and turn it into `TagKeyNotFound`. Same with other method "re-uses."

bretthoerner commented on src/sentry/tagstore/snuba/backend.py:
```diff
@@ -16,69 +17,363 @@
 
     def get_group_tag_keys(self, project_id, group_id, environment_id, limit=None):
-        pass
+        limit = limit or 1000
```
Should this be in the method sig or was it temporary?

alex-hofsteede replied:
fixed

bretthoerner commented on src/sentry/tagstore/snuba/backend.py:
```diff
@@ -16,69 +17,363 @@
+        aggregations = [
+            ['count', '', 'count'],
+            ['min', SEEN_COLUMN, 'first_seen'],
+            ['max', SEEN_COLUMN, 'last_seen'],
```
We apparently use `event.datetime`, that's creation time and not `received` I think: https://github.com/getsentry/sentry/blob/9c87c1ae59f6d1eae39e2bcecee0b192cb8bf3f5/src/sentry/event_manager.py#L1216

alex-hofsteede replied:
fixed

bretthoerner commented on src/sentry/tagstore/snuba/backend.py:
```diff
@@ -16,69 +17,363 @@
 
     def get_group_tag_value_count(self, project_id, group_id, environment_id, key):
-        pass
+        start, end = self.get_time_range(7)
```
nit: We only limit to 7 days in the non-Postgres case currently. On Postgres we limited by rows (for perf/timeouts) though, so I'm not sure what to do here off the top of my head.

alex-hofsteede replied:
Ah yeah, may as well just conform to the postgres version so that the results are somewhat comparable when testing.

bretthoerner commented on src/sentry/tagstore/snuba/backend.py:
```diff
@@ -16,69 +17,363 @@
+                'name': tagstore.get_tag_value_label(key, val),
+                'key': tagstore.get_standardized_key(key),
+                'value': val,
+                # TODO we don't know any of these without more queries
```
Hm, I wonder if they're there just to be serialized "just cause" and are never used?

alex-hofsteede replied:
Will ignore them for now and if they are needed I'll find out.

bretthoerner commented on src/sentry/tagstore/snuba/backend.py:
```diff
@@ -16,69 +17,363 @@
+        }
+        # TODO implement environment_id exclusion, its a little bit more complex
+        # than adding a != condition because environment_ids need to be translated
+        # to filters in snuba.
```
Would it help if we actually passed in `environment` objects to tagstore methods? Starting to regret the "least knowledge" parameters now. I may see about refactoring that.

alex-hofsteede replied:
I think it's more or less ok, with the ids. Its just that the snuba util attempts to translate the columns in `filters` but not `conditions` at the moment. It shouldn't be too hard to add that.

bretthoerner commented on src/sentry/tagstore/snuba/backend.py:
```diff
@@ -16,69 +17,363 @@
+        # than adding a != condition because environment_ids need to be translated
+        # to filters in snuba.
+
+        # TODO OR conditions?
```
Ah crap, is that in the API yet? Forgot we do that here.

alex-hofsteede replied:
Not yet, I'm just trying to think of the nicest way to add it. Do I allow for arbitrary combinations of eg. `foo OR (bar AND baz)` or just go for something simpler but less flexible.

bretthoerner replied:
My hunch is something easy (toplevel `OR` joining a bunch of `AND`s or vice versa) will be enough for a long time. We'll need a lot of product-level changes to actually require more I think?

bretthoerner reviewed (changes requested):
Putting my first-pass notes, mostly small stuff, nice work.

alex-hofsteede committed:
Fixes for review comments

alex-hofsteede committed:
Take advantage of OR support in snuba for user matching

alex-hofsteede committed:
no 7day limit

alex-hofsteede committed:
Mock other tagstore methods

Need to have implementations of these or everything fails with
NotImplementedError.

Return actual (or dummy) results for get_or_create methods, do nothing
for delete/increment/update methods.

_qs methods should probably return something, but its possible this
whole class can just be refactored to not use them

alex-hofsteede committed:
Some more minor compatibility tweaks
